Solver/elastic_law.py

```diff
@@ -0,0 +1,310 @@
+import shenfun as sf
+import sympy as sp
+import numpy as np
+import itertools as it
+from shenfun import inner, div, grad, Dx
+
+
+class LinearCauchyElasticity:
+    def __init__(self):
+        self.name = "LinearCauchyElasticity"
+        self.nb_material_parameters = 2
+
+    def check_with_pde(self, u_hat, material_parameters, body_forces):
+        assert isinstance(u_hat, sf.Function)
+        assert len(material_parameters) == self.nb_material_parameters
+
+        lambd, mu = material_parameters
+        V = u_hat.function_space().get_orthogonal()
+        # left hand side of pde
+        lhs = (lambd + mu) * grad(div(u_hat)) + mu * div(grad(u_hat))
+
+        error_array = sf.Array(V, buffer=body_forces)
+        error_array += sf.project(lhs, V).backward()
+
+        error = np.sqrt(inner((1, 1), error_array ** 2))
+        # scale by magnitude of solution
+        scale = np.sqrt(inner((1, 1), u_hat.backward() ** 2))
+
+        return error / scale
+
+    def compute_body_forces(self, u):
+        assert hasattr(self, "material_parameters")
+        assert len(self.material_parameters) == self.nb_material_parameters
+
+        dim = len(u)
+        lmbda, mu = self.material_parameters
+        x, y, z = sp.symbols("x,y,z")
+        coord = [x, y, z]
+
+        Divergence = 0.
+        for i in range(dim):
+            Divergence += u[i].diff(coord[i])
+
+        GradDiv = np.empty(dim, dtype=sp.Expr)
+        for i in range(dim):
+            GradDiv[i] = Divergence.diff(coord[i])
+
+        Laplace = np.zeros(dim, dtype=sp.Expr)
+        for i in range(dim):
+            for j in range(dim):
+                Laplace[i] += u[i].diff(coord[j], 2)
+
+        body_forces = - (lmbda + mu) * GradDiv - mu * Laplace
+
+        return tuple(body_forces)
+
+    def compute_cauchy_stresses(self, u_hat):
+        assert isinstance(u_hat, sf.Function)
+
+        space = u_hat[0].function_space().get_orthogonal()
+        dim = len(space.bases)
+        N = [u_hat.function_space().spaces[0].bases[i].N for i in range(dim)]
+        lmbda, mu = self.material_parameters
+
+        # displacement gradient
+        H = np.empty(shape=(dim, dim, *N))
+        for i in range(dim):
+            for j in range(dim):
+                H[i, j] = sf.project(Dx(u_hat[i], j), space).backward()
+
+        # linear strain tensor
+        E = 0.5 * (H + np.swapaxes(H, 0, 1))
+
+        # trace of linear strain tensor
+        trE = np.trace(E)
+
+        # create block with identity matrices on diagonal
+        identity = np.zeros_like(H)
+        for i in range(dim):
+            identity[i, i] = np.ones(N)
```

Comment:
`np.ones` return a vector of ones with length `N`. Of course, this is equivalent to an identity matrix but then `identity` contains vectors whereas `E` is a `np.ndarray` of, I guess, `shenfun.Array`.

Reply:
`N` is a tuple or list like `(20, 20)` so this should produce a matrix conatining all ones which has the same shape as the corresponding `shenfun.Array`.

Reply:
Thank you for clarifying this.